acorn/src/scope.js

```diff
@@ -57,12 +59,23 @@ pp.declareName = function(name, bindingType, pos) {
         break
       }
       scope.var.push(name)
+      if (this.inModule && (scope.flags & SCOPE_TOP))
+        delete this.undefinedExports[name];
       if (scope.flags & SCOPE_VAR) break
     }
   }
   if (redeclared) this.raiseRecoverable(pos, `Identifier '${name}' has already been declared`)
 }
 
+pp.checkLocalExport = function(id) {
+  // scope.functions must be empty as Module code is always strict.
+  if (this.scopeStack[0].lexical.indexOf(id.name) === -1 &&
+      this.scopeStack[0].var.indexOf(id.name) === -1 &&
+      !this.undefinedExports[id.name]) {
```

Comment:
Are `id.name`s from `Object.prototype` going to be a problem here? Initialize with `Object.create(null)` instead of `{}`?

Reply:
Hmm indeed. I would usually do that but Acorn currently doesn't use `Object.create` anywhere. I'll use `has` instead of an existence check then.

Reply:
You could also remove the last condition and just always write—the end result would be the same.

Reply:
Indeed! I forgot about the fact that duplicate exports are filtered out.